Fix missing exportables

litewarp opened the pull request:
fixes #95

litewarp committed:
run eslint fix

litewarp committed:
add failing test, bump postgraphile

litewarp committed:
rebuild yarn and refactor PgManyToManyRelationPlugin

litewarp committed:
remove uneeded files

litewarp committed:
do not recreate a schema on export test; add plan names to functions

litewarp commented:
An error occurs when attempting to export the schema because of a Class error --

```
      Error: Attempting to export an instance of a class (at Bar.fields[foosByJunctionJBarIdAndJFooId].extensions.grafast.plan[$$scope][1]); you should wrap this definition in EXPORTABLE! (Class: undefined)
          at _convertToAST (/Users/litewarp/projects/pg-many-to-many/node_modules/graphile-export/src/exportSchema.ts:1039:13)
          at convertToIdentifierViaAST (/Users/litewarp/projects/pg-many-to-many/node_modules/graphile-export/src/exportSchema.ts:1162:5)
          at /Users/litewarp/projects/pg-many-to-many/node_modules/graphile-export/src/exportSchema.ts:1279:12
          at Array.map (<anonymous>)
          at factoryAst (/Users/litewarp/projects/pg-many-to-many/node_modules/graphile-export/src/exportSchema.ts:1265:37)
          at importWellKnownOrFactory (/Users/litewarp/projects/pg-many-to-many/node_modules/graphile-export/src/exportSchema.ts:1137:12)
          at convertToIdentifierViaAST (/Users/litewarp/projects/pg-many-to-many/node_modules/graphile-export/src/exportSchema.ts:1161:5)
          at /Users/litewarp/projects/pg-many-to-many/node_modules/graphile-export/src/exportSchema.ts:1595:13
          at Array.forEach (<anonymous>)
          at exportSchemaTypeDefs (/Users/litewarp/projects/pg-many-to-many/node_modules/graphile-export/src/exportSchema.ts:1574:15)
          at exportSchemaAsString (/Users/litewarp/projects/pg-many-to-many/node_modules/graphile-export/src/exportSchema.ts:2048:5)
          at exportSchema (/Users/litewarp/projects/pg-many-to-many/node_modules/graphile-export/src/exportSchema.ts:2177:26)
          at exportSchema (/Users/litewarp/projects/pg-many-to-many/__tests__/integration/export/export.test.js:50:13)
          at processTicksAndRejections (node:internal/process/task_queues:105:5)
          at withPgClient (/Users/litewarp/projects/pg-many-to-many/__tests__/helpers.js:22:20)
```

Looks like the error results from [this check]( https://github.com/graphile/crystal/blob/990e7ebcb82f319d3caadef635c7a00cdce7ee5e/utils/graphile-export/src/exportSchema.ts#L1036)

benjie commented:
> Error: Attempting to export an instance of a class (at Bar.fields[foosByJunctionJBarIdAndJFooId].extensions.grafast.plan[$$scope][1]); you should wrap this definition in EXPORTABLE! (Class: undefined)

This is saying that index `1` of the `plan` `$$scope` (i.e. the 2nd argument to the `EXPORTABLE` function that the `plan` is wrapped with) is not EXPORTABLE. It's detecting this badly because it's going through the fallback code, I'll update the code you link to this:

```ts
      if (thing.constructor) {
        throw new Error(
          `Attempting to export an instance of a class (at ${locationHint}); you should wrap this definition in EXPORTABLE! (Class: ${thing.constructor})`,
        );
      } else {
        throw new Error(
          `Attempting to export non-POJO object (at ${locationHint}); you should wrap this definition in EXPORTABLE! (Prototype: ${prototype})`,
        );
      }
```

But the error remains. Looking over your changes, the second argument to plans include `sql.identifier`, which should be caught by the sql.isSQL check, but if you have two versions of `pg-sql2` installed then this may not work... or... Yeah, I found it. The fallback code for the isSQL check was too rigid:

https://github.com/graphile/crystal/blob/990e7ebcb82f319d3caadef635c7a00cdce7ee5e/utils/graphile-export/src/exportSchema.ts#L68-L81

Specifically the symbol description is now `pg-sql2-type-${revision}` so this match will fail. That explains why the error message is confusing, but what you care about is how to actually fix the issue. This is fortunately fairly simple: track down all the second arguments to plan, for example, `junctionAlias`, and make sure that they are themselves exportable. More specifically, this line:

https://github.com/litewarp/pg-many-to-many/blob/52c30adcbdc6100d1f464fa1c922bd5d80c99668/src/PgManyToManyRelationPlugin.ts#L201

needs to become:

```ts
                const junctionAlias = EXPORTABLE((sql, junctionSymbol) => sql.identifier(junctionSymbol), [sql, junctionSymbol]);
```

Though I don't think symbols themselves are exportable... We normally have to handle this via Graphile Export's "well known" system - i.e. you need to import a module to get the symbol, but we don't want this for this plugin (don't want to import it at runtime). We can't store info onto symbols:

```
> const s = Symbol('foo')
undefined
> s.bar = 1
1
> s.bar
undefined
```

so we need to wrap it.

Change line 12:

https://github.com/litewarp/pg-many-to-many/blob/52c30adcbdc6100d1f464fa1c922bd5d80c99668/src/PgManyToManyRelationPlugin.ts#L12

to:

```
export const junctionSymbolContainer = EXPORTABLE(() => ({symbol: Symbol("junction")}), []);
```

and then update all references to `junctionSymbol` to instead reference `junctionSymbolContainer.symbol`. That should solve it.

benjie commented:
The error message issues are resolved in https://github.com/graphile/crystal/pull/2497, but the fixes you need to make remain the same.

litewarp commented:
> 
> Change line 12:
> 
> https://github.com/litewarp/pg-many-to-many/blob/52c30adcbdc6100d1f464fa1c922bd5d80c99668/src/PgManyToManyRelationPlugin.ts#L12
> 
> to:
> 
> ```
> export const junctionSymbolContainer = EXPORTABLE(() => ({symbol: Symbol("junction")}), []);
> ```
> 
> and then update all references to `junctionSymbol` to instead reference `junctionSymbolContainer.symbol`. That should solve it.

So I changed the followings lines and updated references 
```ts
// line 12
export const junctionSymbolContainer = EXPORTABLE(() => ({symbol: Symbol("junction")}), []);

// line 201
const junctionAlias = EXPORTABLE((sql, junctionSymbolContainer) => sql.identifier(junctionSymbolContainer.symbol), [sql, junctionSymbolContainer]);
```

But now I'm getting the following error:

```
console.warn
    Recoverable error occurred; use envvar 'DEBUG="graphile-build:warn"' for full error (see: https://postgraphile.org/postgraphile/current/debugging )
    > TypeError: Cannot define property $exporter$args, object is not extensible…

      151 |           relationships.reduce(
      152 |             (memo, relationship) =>
    > 153 |               build.recoverable(memo, () => {
          |                     ^
      154 |                 const {
      155 |                   leftTable,
      156 |                   leftRelationName,

      at Object.swallowError (node_modules/graphile-build/src/swallowError.ts:24:15)
      at Object.recoverable (node_modules/graphile-build/src/makeNewBuild.ts:178:14)
      at recoverable (src/PgManyToManyRelationPlugin.ts:153:21)
          at Array.reduce (<anonymous>)
      at reduce (src/PgManyToManyRelationPlugin.ts:151:25)
      at SchemaBuilder.applyHooks (node_modules/graphile-build/src/SchemaBuilder.ts:161:20)
      at Object.fields (node_modules/graphile-build/src/newWithHooks/index.ts:305:42)
      at fields (node_modules/grafast/dist/utils.js:282:69)
      at resolveObjMapThunk (node_modules/graphql/type/definition.js:504:40)
      at defineFieldMap (node_modules/graphql/type/definition.js:766:20)
      at GraphQLObjectType._fields (node_modules/graphql/type/definition.js:691:26)
      at GraphQLObjectType.getFields (node_modules/graphql/type/definition.js:710:27)
      at Object.getTypeByName (node_modules/graphile-build/src/makeNewBuild.ts:415:36)
      at node_modules/graphile-build-pg/src/plugins/PgRelationsPlugin.ts:1449:29
          at Array.reduce (<anonymous>)
      at addRelations (node_modules/graphile-build-pg/src/plugins/PgRelationsPlugin.ts:1424:18)
      at SchemaBuilder.applyHooks (node_modules/graphile-build/src/SchemaBuilder.ts:161:20)
      at Object.fields (node_modules/graphile-build/src/newWithHooks/index.ts:305:42)
      at fields (node_modules/grafast/dist/utils.js:282:69)
      at resolveObjMapThunk (node_modules/graphql/type/definition.js:504:40)
      at defineFieldMap (node_modules/graphql/type/definition.js:766:20)
      at GraphQLObjectType._fields (node_modules/graphql/type/definition.js:691:26)
      at GraphQLObjectType.getFields (node_modules/graphql/type/definition.js:710:27)
      at Object.getTypeByName (node_modules/graphile-build/src/makeNewBuild.ts:415:36)
      at node_modules/graphile-build-pg/src/plugins/PgRowByUniquePlugin.ts:120:38
          at Array.reduce (<anonymous>)
      at node_modules/graphile-build-pg/src/plugins/PgRowByUniquePlugin.ts:105:59
      at Object.recoverable (node_modules/graphile-build/src/makeNewBuild.ts:176:16)
      at node_modules/graphile-build-pg/src/plugins/PgRowByUniquePlugin.ts:104:19
          at Array.reduce (<anonymous>)
      at GraphQLObjectType_fields (node_modules/graphile-build-pg/src/plugins/PgRowByUniquePlugin.ts:102:26)
      at SchemaBuilder.applyHooks (node_modules/graphile-build/src/SchemaBuilder.ts:161:20)
      at Object.fields (node_modules/graphile-build/src/newWithHooks/index.ts:305:42)
      at fields (node_modules/grafast/dist/utils.js:282:69)
      at resolveObjMapThunk (node_modules/graphql/type/definition.js:504:40)
      at defineFieldMap (node_modules/graphql/type/definition.js:766:20)
      at GraphQLObjectType._fields (node_modules/graphql/type/definition.js:691:26)
      at GraphQLObjectType.getFields (node_modules/graphql/type/definition.js:710:27)
      at Object.getTypeByName (node_modules/graphile-build/src/makeNewBuild.ts:415:36)
      at callback (node_modules/graphile-build/src/plugins/QueryPlugin.ts:67:25)
      at SchemaBuilder.applyHooks (node_modules/graphile-build/src/SchemaBuilder.ts:161:20)
      at node_modules/graphile-build/src/newWithHooks/index.ts:117:37
      at SchemaBuilder.newWithHooks (node_modules/graphile-build/src/newWithHooks/index.ts:862:7)
      at SchemaBuilder.buildSchema (node_modules/graphile-build/src/SchemaBuilder.ts:265:29)
      at buildSchema (node_modules/graphile-build/src/index.ts:485:26)
      at make (node_modules/graphile-build/src/index.ts:573:31)
      at __tests__/integration/export/export.test.js:22:24
      at withPgClient (__tests__/helpers.js:22:20)
```

benjie commented:
Something that you're trying to make exportable is `Object.freeze(...)`'d or similar.

benjie commented:
Could be the SQL actually; if so you'll need to do a similar trick as with the symbol (i.e. wrap it in a mutable object); have a look and see if we do anything similar in PostGraphile.

litewarp commented:
Would this be a [good example](https://github.com/graphile/crystal/blob/990e7ebcb82f319d3caadef635c7a00cdce7ee5e/graphile-build/graphile-build-pg/src/plugins/PgRelationsPlugin.ts#L32)? -- it looks like you created refs using tamedevil and then wrap it again

```ts
// lines 32-33
const ref_connection = te.ref(connection, "connection");
const ref_sql = te.ref(sql, "sql");

// lines 1182 - 1190
remoteAttributes.forEach((remoteAttrName, i) => {
  functionLines.push(te`  ${collectionIdentifier}.where(${ref_sql}\`\${previousAlias}.\${${ref_sql}.identifier(${te.lit(
                    remoteAttrName,
  )})} = \${${collectionIdentifier}.placeholder($record.get(${te.lit(
                    localAttributes[i],
                  )}))}\`);`,
  );
});

```

litewarp committed:
it exports without error

litewarp commented:
@benjie it exports according to the basic test without error. Let me know what other kinds of work you want me to do (e.g., a snapshot test on the export?). Thanks for all your guidance.

benjie commented on src/PgManyToManyRelationPlugin.ts:
```diff
@@ -1,15 +1,18 @@
 
-export const junctionSymbol = Symbol("junction");
+export const junctionSymbolContainer = EXPORTABLE(
+  () => Object.assign({}, { symbol: Symbol("junction") }),
```
I don't think there's any benefit to using Object.assign here?
```suggestion
  () => ({ symbol: Symbol("junction") }),
```

benjie committed:
Simplify changes to yarn.lock

benjie reviewed (approved):
Code changes all look good to me :+1:

benjie commented:
Released as  @graphile-contrib/pg-many-to-many@2.0.0-beta.8
